cpp/src/arrow/csv/converter.cc

```diff
@@ -390,23 +390,45 @@ class TimestampConverter : public ConcreteConverter {
     using value_type = TimestampType::c_type;
 
     TimestampBuilder builder(type_, pool_);
-    StringConverter<TimestampType> converter(type_);
+    auto& converters = options_.timestamp_converters;
 
-    auto visit = [&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
-      value_type value = 0;
-      if (IsNull(data, size, quoted)) {
-        builder.UnsafeAppendNull();
-        return Status::OK();
-      }
-      if (ARROW_PREDICT_FALSE(
-              !converter(reinterpret_cast<const char*>(data), size, &value))) {
+    if (!converters.empty()) {
+      auto visit = [&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
+        for (auto& converter : converters) {
+          value_type value = 0;
+          if (IsNull(data, size, quoted)) {
+            builder.UnsafeAppendNull();
+            return Status::OK();
+          }
+          if (!converter(type_, reinterpret_cast<const char*>(data), size, &value)) {
+            continue;
+          }
+          builder.UnsafeAppend(value);
+          return Status::OK();
+        }
         return GenericConversionError(type_, data, size);
-      }
-      builder.UnsafeAppend(value);
-      return Status::OK();
-    };
-    RETURN_NOT_OK(builder.Resize(parser.num_rows()));
-    RETURN_NOT_OK(parser.VisitColumn(col_index, visit));
+      };
+      RETURN_NOT_OK(builder.Resize(parser.num_rows()));
+      RETURN_NOT_OK(parser.VisitColumn(col_index, visit));
+    } else {
+      StringConverter<TimestampType> converter(type_);
+
+      auto visit = [&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
+        value_type value = 0;
+        if (IsNull(data, size, quoted)) {
+          builder.UnsafeAppendNull();
+          return Status::OK();
+        }
+        if (ARROW_PREDICT_FALSE(
+                !converter(reinterpret_cast<const char*>(data), size, &value))) {
+          return GenericConversionError(type_, data, size);
+        }
+        builder.UnsafeAppend(value);
+        return Status::OK();
+      };
+      RETURN_NOT_OK(builder.Resize(parser.num_rows()));
+      RETURN_NOT_OK(parser.VisitColumn(col_index, visit));
```

Comment:
Rather than duplicate code, create an ISO-8601 parser implementation and instantiate that as the default. Need to check the performance impact of changing the inline conversion to a virtual call